MSC: Support for multi-stream VoIP (#3077)
matrix-org/matrix-doc: proposals/3077-multi-stream-voip.md

`id` in the session description. The values are of `SDPStreamMetadata` type and
have the following fields:

+ `purpose` - a string indicating the purpose of the stream. For compatibility
  between clients values `m.usermedia` and `m.screenshare` are defined.
  `m.usermedia` is the stream that contains the webcam and microphone tracks.
  `m.screenshare` is then the stream with the screen-sharing tracks.

### Example

```JSON
{
    "type": "m.call.invite",
    "room_id": "!roomId",
    "content": {
        "call_id": "1414213562373095",
        "invitee": "@bob:matrix.org",
        "party_id": "1732050807568877",
        "lifetime": "60000",
        "capabilities": {
            "m.call.transferee": true,
        },
        "offer": {
            "sdp": "...",
            "type": "offer",
        },
        "sdp_stream_metadata": {
            "271828182845": {
                "purpose": "m.screenshare",
            },
            "314159265358": {
                "purpose": "m.usermedia",
            },
        },
        "version": "1",
    },
}
```

### Edge cases

+ There should _never_ be more than one `SDPStreamMetadata` in

>   `sdp_stream_metadata` with the same `purpose`.

Reviewer: Why not? I can easily see the desire to share two windows or similar.

Author: That would make the problem appear again - there would be two streams with the same description and no way to differentiate them. I assumed that what you want isn't very common, so I assumed that if a client wanted this it could set `purpose` to some custom string (e.g. `screenshare1`, `screenshare2`).

I could extend this MSC with something like `name` and that could be the thing that should never repeat

Reviewer: `screenshare2` isn't really a purpose, it is just a second screenshare. I think that if we can't represent this we should probably make it a bit more flexible. They are both just screenshares. This would still solve some of the problems the MSC aimed for without adding a huge limitation as now as you know that they are both screenshares and you know who they are from.

It seems that the only purpose that we miss out on is replacement, but I am a little confused by this case anyways. Is the idea that we can have a soft-handover where you can send both streams until all clients transition then you drop the old one? Why is a replacement not just dropping the old stream? If it is more than that then I think it would make sense to have explicit metadata about replacements such as a `replaces` field.

Author: Yeah, you're right. Right now I feel like a `replaces` field makes the most sense, but I'll think about this some more. I also wonder if @dbkr has any thoughts about this

Reviewer 2: So my reading of what Simon was saying is that if you sent multiple screenshares, for example, you'd probably want to add some information to differentiate them which would be more spec on top of this, but you'd also want clients that only understood this MSC and not that new bit of spec to display the same one, consistently - and probably a specific one. To that end, I'd argue that `screenshare2` *is* a purpose: it's the auxiliary screenshare as opposed to the main one, ie. if you add another one as `screenshare2` then clients that don't understand that will display the main one and ignore the extra one, which is good.

Related question: should we use the same mechanism for signalling front/rear facing camera feeds? Should we allow sending both front-facing and rear-facing at the same time, and if we do, what should a client do if it can only display one? (Also, would we need an extra event type to update a stream's metadata without any changes to the stream(s)?)

Author: > should we use the same mechanism for signalling front/rear facing camera feeds? Should we allow sending both front-facing and rear-facing at the same time

For switching front and rear `replaceTrack()` probably makes the most sense. Although we could allow sending both. But I 
am not sure how often would this get used. 

> Also, would we need an extra event type to update a stream's metadata without any changes to the stream(s)?

Probably yes. We could send `m.call.negotiate` with no SDP changes but that seems ugly. I am not sure why would we want to even do this though? (I guess muting as discussed in [#16862](https://github.com/vector-im/element-web/issues/16862) could be a use case but I feel like `m.call.mute` would be better for that)

> what should a client do if it can only display one?

Yeah, good question. Should we support that? Should we set something like a main stream?

Some parts of this come down to how complex scenarios should we allow by default (multiple cams, screenshares, other things?). I don't think I know the answer to this. :man_shrugging: 

Anyway, having a `replaces` field might not be a bad idea in general? It would theoretically make stream replacements easier to implement

Author: Should I split this into separate threads?

Reviewer 2: I'm not sure I follow what `replaces` is trying to achieve. Wouldn't you just keep the ID of the stream the same? If having an overlap were a concern, you could use a new SSRC and continue to send the old stream with old SSRC for a bit?

and sorry, yep - I should have made a separate thread for camera identification: will do

Author: > Wouldn't you just keep the ID of the stream the same?

Can we do that though? Since we can't set the ID (at least not on the web), we'd have to replace tracks which I am not sure is the best solution? Perhaps the MSC should codify that the stream ID wouldn't change in the case of replacement?

Reviewer: >  I'd argue that screenshare2 is a purpose: it's the auxiliary screenshare as opposed to the main one

Who said the first one was a main one? They can be just as important. I think if we want some stream to be higher priotity we should signal that explicitly, rather than adding `m.usermedia2` and `m.screenshare3`. Purpose doesn't feel like the right place to express priority.

Author: My original aim was to support the typical situation of today's most (if not all) VoIP apps - one `m.usermedia` and one `m.screenshare` stream. I thought that support for more than one would be done using a special `purpose` as this would be app-specific. 

I think this can be split into:

1. Should we add a `replaces`? Can this work without it if we allow multiple streams of the same `purpose`?
2. How complex should this be for the average app? In my mind, if an app implements multi-stream support, it shouldn't be that much of a problem to support any number of streams - this shouldn't be a problem in the end if we add `replaces`.
3. Should we support identifying the "main" stream and stream priority? Does that have any benefits? Should it be a part of this MSC?

Do you agree to split the thread into these three?

+ If an incoming stream is not described, in `sdp_stream_metadata`, it should be

>   ignored.

Reviewer: How does this work during the transition period. The new clients won't see a `sdp_stream_metadata` field. Should the old behaviour be explicitly specified in that case.

Author: My thinking was that during the initial `invite` and `answer` the client will figure out if the opponent supports this based on whether or not `sdp_stream_metadata` is present in the opponent's message. If it isn't they should behave the way they do now. I've tried to describe this in _Backwards compatibility_ but maybe I wasn't very clear?

Reviewer: Ah, I think I just missed that section. Sorry.

Author: Ah, I'll try to make it more clear anyway

  `m.screenshare`), it should be ignored.

### Backwards compatibility

If the field `sdp_stream_metadata` is missing, clients should ignore any new
incoming streams (i.e. they should use the first one).

## Alternatives

Setting the stream `id`s to custom values had been considered. Though this is
possible on some platforms, it is not in browsers. That is because the `id`
property of `MediaStream` is _read-only_ as the [MDN Documentation
states](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream/id).
Similar is true for SDP attributes.

This proposal is also more practical for cases where more complex metadata is
needed. For conferencing, a `user_id` field could be added to
`SDPStreamMetadata`.

## Unstable prefix

During development, the following fields should be used:

|Release                     |Development                                    |
|----------------------------|-----------------------------------------------|
|`sdp_stream_metadata`       |`org.matrix.msc3077.sdp_stream_metadata`       |
|`m.call.sdp_stream_metadata`|`org.matrix.msc3077.call.sdp_stream_metadata`|

## Potential issues

None that I can think of.

> ## Security considerations

Reviewer: Earlier you say that

> if conferencing using an SFU there would be no way to know from which user a stream is coming from

It seems like this proposal probabilistically solves this issue but does not securely solve this issue. I don't think there is anything preventing a race where a user sees a stream ID (either via a faster-syncing homeserver or seeing it in the call before the event arrives) and spoofing an event claiming that those streams are from them.

Author: My idea is that a client calls an SFU and they exchange data as if they were two users in a 1:1 call except there is a lot of streams involved. This way the SFU forwards the metadata. Where would the spoofing happen here?

What I think you're suggesting is that the client gets the data directly from other call participants. In that case, something like this could probably happen.

I don't know if my idea of how would an SFU work makes sense, so I may be completely wrong here

Reviewer: I see. It wasn't clear to me that this proposal required a SFU. Maybe that should made more clear?

But in that case we still have the issue that we are trusting the SFU to tell us who the stream is from, but I guess that isn't a new problem.

Author: > It wasn't clear to me that this proposal required an SFU

This MSC does _not_  require an SFU. At the moment this is just for 1:1 calls. Conferencing is a separate thing. I mentioned it because it felt like a possible use case but it isn't the focus of this MSC.

Reviewer: Ah ok. For 1:1 calls this makes sense.

screen-sharing.

The simplest possible implementation of VoIP call upgrades would assume that any
new incoming stream would replace the old one. This solution works, but problems
start to appear with the introduction of screen-sharing. There is no way to know
which one of the already existing incoming streams (user-media and
screen-sharing) should a new incoming stream replace. Further, if conferencing
using an SFU (as suggested in
[MSC2359](https://github.com/matrix-org/matrix-doc/pull/2359)) were implemented
there would be no way to know from which user a stream is coming from.
Therefore, a way to differentiate streams from each other is needed.

## Proposal

This MSC proposes adding an `sdp_stream_metadata` field to the events containing
a session description i.e.:

+ `m.call.invite`
+ `m.call.answer`
+ `m.call.negotiate`

The `sdp_stream_metadata` field is an object in which each key is one stream
`id` in the session description. The values are of `SDPStreamMetadata` type and
have the following fields:

+ `purpose` - a string indicating the purpose of the stream. For compatibility
  between clients values `m.usermedia` and `m.screenshare` are defined.

>   `m.usermedia` is the stream that contains the webcam and microphone tracks.

Reviewer 2: continuing thread on whether we should identify front/rear facing cameras with this, for purposes like when the user rotates the camera and the same stream changes to a different camera (eg. FaceTime displays an animation on the remote side to show that the camera has changed). I feel like this is probably something for separate MSC, unless we want two separate purposes for front/rear camera.

Author: Yeah, unless we want a separate `purpose` for that probably a separate MSC. If we wanted a separate `purpose`, would we have `m.usermedia`, `m.usermedia.front` and `m.usermedia.rear` or something else?

Author: > eg. FaceTime displays an animation on the remote side to show that the camera has changed

This bit seems more like a stream replacement problem rather than a `purpose` problem. But... maybe we could allow sending both and set a main stream which would tell the client on the other side to do the animation

Author: My suggestion would be adding something like `camere_type` to further distinguish the cameras - this way not all clients need to support this and we avoid having two `purpose`s that are almost the same

>   `m.screenshare` is then the stream with the screen-sharing tracks.

Author: Perhaps we should add another field `replaces`, that would explicitly specify what stream to replace in case of a stream replacement. If we allow only one stream of each purpose, this wouldn't be necessary, but that might not be ideal. We also might be able to use the same stream id (but I am not sure we can do that everywhere).

Author: See https://github.com/matrix-org/matrix-spec-proposals/pull/3077#discussion_r915648792

### Example

```JSON
{
    "type": "m.call.invite",
    "room_id": "!roomId",
    "content": {
        "call_id": "1414213562373095",
        "invitee": "@bob:matrix.org",
        "party_id": "1732050807568877",
        "lifetime": "60000",
        "capabilities": {
            "m.call.transferee": true,
        },
        "offer": {
            "sdp": "...",
            "type": "offer",
        },
        "sdp_stream_metadata": {
            "271828182845": {
                "purpose": "m.screenshare",
            },
            "314159265358": {
                "purpose": "m.usermedia",
            },
        },
        "version": "1",
    },
}
```

### Edge cases

+ There should _never_ be more than one `SDPStreamMetadata` in

>   `sdp_stream_metadata` with the same `purpose`.

Author: A question that has arisen in other threads is how complex should this be - should we allow multiple streams of the same `purpose`, or should we keep that to be app-specific (as this isn't a thing we would usually expect in a standard messenger)?

In my mind, if an app implements multi-stream support, it shouldn't be that much of a problem to support any number of streams - this shouldn't be a problem in the end if we add `replaces`.

Reviewer: (message moved here because github email apparently doesn't respect threads)

I think that sounds good. In general I think it is fine to write the spec
with a single stream of each in mind. However Matrix is about more than
just video conferences and if we say [0 1] it is hard to change in the
future. I think if we say any number now, it won't be hard for most
clients, we can even say "clients supporting only one stream of a given
type should ignore all but the first of the type" or something link that.

As for replaces and priority I think we should defer them. Both to keep
this simpler and give time to collect more specific use cases. For now just
dropping and old stream should be "good enough" for replaces and if clients
are only sending one stream of each type we don't really need priority (we
could have it between usermedia and screen share but for now clients can
just do something reasonable).

I think it makes sense to keep this MSC simple, but I don't want to limit
our options in the future without making a while separate protocol for the
more complex cases.

Reviewer 3: One use case that is similar to the rear/front camera issue brought up by @dbkr would be the ability to share multiple monitors at the same time.
Therefore I think it should definitely be possible that clients send multiple streams with the same base purpose and that should be noted in the spec. For clients that do not support this we could then define the convention proposed by @kevincox that clients may choose to display only  the first stream of a supported purpose.
For handling the additional metadata that clients might want to add to differentiate between streams of the same purpose we could add a new stream tags map:
```json
{
    ...
    "sdp_stream_metadata": {
      "271828182845": {
        "purpose": "m.screenshare",
        "tags": {
          "displayPositionX": 0,
          "displayPositionY": 0
        }
      },
      "371828182845": {
        "purpose": "m.screenshare",
        "tags": {
          "displayPositionX": 1,
          "displayPositionY": 0
        }
      },
      "414159265358": {
        "purpose": "m.usermedia",
        "tags": {
          "type": "camera",
          "direction": "front"
        }
      },
      "514159265358": {
        "purpose": "m.usermedia",
        "tags": {
          "type": "camera",
          "direction": "back"
        }
      }
    },
    "version": "1"
  }
}
```
This would allow clients to add additional metadata to the streams until that is further standardized in a follow-up MSC.

Author: I think the most straightforward and most future-proof method is to add an `id` field which ensures streams have a truly unique identifier

Author: Okay, so after doing more research and testing, it turns out the situations might be simpler than previously thought: stream-replacement is a thing that doesn't ever really seem to happen, so having `purpose` (or another field) be unique seems rather unnecessary

Could people please confirm this really is the case for every implementation?

During the initial invite and answer exchange clients find out if the field
`sdp_stream_metadata` is missing. If it is not present in the event sent by the
opponent, the client should ignore any new incoming streams (i.e. it should use
the first one) and it shouldn't send more than one stream.

## Alternatives

Setting the stream `id`s to custom values had been considered. Though this is
possible on some platforms, it is not in browsers. That is because the `id`
property of `MediaStream` is _read-only_ as the [MDN Documentation
states](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream/id).
Similar is true for SDP attributes.

This proposal is also more practical for cases where more complex metadata is
needed. For conferencing, a `user_id` field could be added to

> `SDPStreamMetadata`; for differentiating between the front and rear camera of a

Reviewer 4: ad725d2 removed the definition of `SDPStreamMetadata`

```suggestion
the objects in `sdp_stream_metadata`; for differentiating between the front and rear camera of a
```

property of `MediaStream` is _read-only_ as the [MDN Documentation
states](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream/id).
Similar is true for SDP attributes.

This proposal is also more practical for cases where more complex metadata is
needed. For conferencing, a `user_id` field could be added to
`SDPStreamMetadata`.

## Unstable prefix

During development, the following fields should be used:

|Release                     |Development                                    |
|----------------------------|-----------------------------------------------|
|`sdp_stream_metadata`       |`org.matrix.msc3077.sdp_stream_metadata`       |

> |`m.call.sdp_stream_metadata`|`m.call.org.matrix.msc3077.sdp_stream_metadata`|

Reviewer 5: It's not very prefixy to add the unstable part at the end

```suggestion
|`m.call.sdp_stream_metadata`|`org.matrix.msc3077.call.sdp_stream_metadata`|
```

Author: I do wonder why did I do that :sweat_smile:

# MSC3077: Support for multi-stream VoIP

This MSC proposes a method for differentiating WebRTC streams from each other.

[MSC2746](https://github.com/matrix-org/matrix-doc/pull/2746) has improved VoIP
immeasurably. Yet, there is still no clear way to handle things such as
screen-sharing.

Simple VoIP calls only ever feature one stream, though often clients will want
to send multiple - usermedia, screensharing and possibly more. In a situation
with more streams, it can be very helpful to provide the other side with
metadata about the content of the streams.

## Proposal

This MSC proposes adding an `sdp_stream_metadata` field to the events containing
a session description i.e.:

+ `m.call.invite`
+ `m.call.answer`

> + `m.call.negotiate`

Reviewer 4: it's very helpful for readers of your proposal if you can link to existing parts of the spec (https://spec.matrix.org/v1.7/client-server-api/#mcallinvite etc): it helps put your proposed changes in context.

The `sdp_stream_metadata` field is an object in which each key is one stream

> `id` in the session description. The values are of `SDPStreamMetadata` type and

Reviewer 4: > are of `SDPStreamMetadata` type

the way this is written makes me think that I should already know what an `SDPStreamMetadata` is.

Unless you are specifically referring to an existing definition (in which case, link it please!), I would remove the `SDPStreamMetadata` name (and just refer to "the `sdp_stream_metadata` objects" or something where you mention it below).

have the following fields:

+ `purpose` - an optional string indicating the purpose of the stream. For compatibility
  between clients values `m.usermedia` and `m.screenshare` are defined.
  `m.usermedia` is the stream that contains the webcam and/or microphone tracks.
  `m.screenshare` is then the stream with the screen-sharing tracks. Multiple

>   media streams may have the same `purpose`.

Reviewer 4: Note that in the rendered version, the linebreaks here are lost. It's not a big problem as it's still clear enough, but consider a nested list?

### Example

```JSON
{
    "type": "m.call.invite",
    "room_id": "!roomId",
    "content": {
        "call_id": "1414213562373095",
        "invitee": "@bob:matrix.org",
        "party_id": "1732050807568877",
        "lifetime": "60000",
        "capabilities": {
            "m.call.transferee": true,
        },
        "offer": {
            "sdp": "...",
            "type": "offer",
        },
        "sdp_stream_metadata": {
            "271828182845": {
                "purpose": "m.screenshare",
            },
            "314159265358": {
                "purpose": "m.usermedia",
            },
        },
        "version": "1",
    },
}
```

### Edge cases

+ If an incoming stream is not described, in `sdp_stream_metadata`, it should be

>   ignored.

Reviewer 4: Presumably this only happens if `sdp_stream_metadata` is actually present? Otherwise we break compatibility with existing clients?

+ If a stream has a `purpose` of an unknown type (i.e. not `m.usermedia` or
  `m.screenshare`), it should be ignored.

### Backwards compatibility

During the initial invite and answer exchange clients find out if the field
`sdp_stream_metadata` is missing. If it is not present in the event sent by the
opponent, the client should ignore any new incoming streams (i.e. it should use

> the first one) and it shouldn't send more than one stream.

Reviewer 4: > it shouldn't send more than one stream.

... ie, if someone wants to do a screenshare, the client replaces the camera feed with the screenshare? Out of interest: is this the way that current implementations behave?

Author: Yep

Reviewer 4: Would be nice just to clarify this.

```suggestion
the first one) and it shouldn't send more than one stream (i.e. clients cannot send a video feed and a screenshare at the same time, as is the case in current clients).
```

        "lifetime": "60000",
        "capabilities": {
            "m.call.transferee": true,
            "m.call.sdp_stream_metadata": true,
        },
        "offer": {
            "sdp": "...",
            "type": "offer",
        },
        "sdp_stream_metadata": {
            "271828182845": {
                "purpose": "m.screenshare",
            },
            "314159265358": {
                "purpose": "m.usermedia",
            },
        },
        "version": "1",
    },
}
```

### Edge cases

+ There should _never_ be more than one `SDPStreamMetadata` in
  `sdp_stream_metadata` with the same `purpose`.
+ If an incoming stream is not described, in `sdp_stream_metadata`, it should be
  ignored.
+ If a stream has `purpose` of an unknown type (i.e. not `m.usermedia` or
  `m.screenshare`), it should be ignored.

### Backwards compatibility and capability advertisement

For backwards compatibility and capability advertisement, a field
`m.call.sdp_stream_metadata` is added to the `capabilities` object. If
`m.call.sdp_stream_metadata` is anything other than `true`, clients should
behave the way they do right now (i.e. they should not allow usage of multiple

> streams).

Reviewer 2: I would be tempted to skip the capability advertisement for this and just say that the absence of stream metadata means clients just take the first or whatever MSC2746 says. I can't think of anything the capability advertisement caters for specifically?

Author: Yeah, true, will do

This MSC proposes adding an `sdp_stream_metadata` field to the events containing
a session description i.e.:

+ [`m.call.invite`](https://spec.matrix.org/unstable/client-server-api/#mcallinvite)
+ [`m.call.answer`](https://spec.matrix.org/unstable/client-server-api/#mcallanswer)
+ [`m.call.negotiate`](https://spec.matrix.org/unstable/client-server-api/#mcallnegotiate)

The `sdp_stream_metadata` field is an object in which each key is one stream

> `id` in the session description. The `sdp_stream_metadata` objects have the
> following fields:

Reviewer 4: ```suggestion
`id` in the session description. The values are objects with the
following fields:
```

### Edge cases

+ If an incoming stream is not described in `sdp_stream_metadata` and
  `sdp_stream_metadata` is present, the stream should be ignored.
+ If a stream has a `purpose` of an unknown type (i.e. not `m.usermedia` or

>   `m.screenshare`) or is missing completely, it should be ignored.

Author: None :P 

```suggestion
  `m.screenshare`), it should be ignored.
```
